gitpod-io/gitpod
# [ws-daemon] Refactor enforcing of CPU limits

## Description

- Remove `jiffies` concept
- Rely on `cpu.cfs_period_us`

## How to test

- Open a workspace
- Download cpuburn `wget https://cdn.pmylund.com/files/tools/cpuburn/linux/cpuburn-1.0-amd64.tar.gz && tar zxpvf cpuburn-1.0-amd64.tar.gz`
- Start CPU burn `./cpuburn/cpuburn`
- After 30s check ws-daemon logs
- The log should contain an output similar to the next one

```
{"containerID":"0f9f1e5553d6510e9a65376976760f3f33f6dffd7cee8d22dc516f5363f0741c","currentQuota":-1,"instanceId":"974f65d2-9e4d-4c75-8825-126e8811605b","level":"info","limit":500,"message":"set new CPU limit","quota":500000,"serviceContext":{"service":"ws-daemon","version":"commit-ee12972d9d496abec7d93a48989384ce8f05338f"},"severity":"INFO","time":"2022-01-24T23:11:12Z","userId":"","workspaceId":"gitpodio-templatetypescr-z3dr46y2j9j"}
{"containerID":"0f9f1e5553d6510e9a65376976760f3f33f6dffd7cee8d22dc516f5363f0741c","currentQuota":500000,"instanceId":"974f65d2-9e4d-4c75-8825-126e8811605b","level":"info","limit":400,"message":"set new CPU limit","quota":400000,"serviceContext":{"service":"ws-daemon","version":"commit-ee12972d9d496abec7d93a48989384ce8f05338f"},"severity":"INFO","time":"2022-01-24T23:11:42Z","userId":"","workspaceId":"gitpodio-templatetypescr-z3dr46y2j9j"}
{"containerID":"0f9f1e5553d6510e9a65376976760f3f33f6dffd7cee8d22dc516f5363f0741c","currentQuota":400000,"instanceId":"974f65d2-9e4d-4c75-8825-126e8811605b","level":"info","limit":200,"message":"set new CPU limit","quota":200000,"serviceContext":{"service":"ws-daemon","version":"commit-ee12972d9d496abec7d93a48989384ce8f05338f"},"severity":"INFO","time":"2022-01-24T23:12:12Z","userId":"","workspaceId":"gitpodio-templatetypescr-z3dr46y2j9j"}
{"containerID":"0f9f1e5553d6510e9a65376976760f3f33f6dffd7cee8d22dc516f5363f0741c","instanceId":"3214228e-a0eb-4b6a-91eb-c34bb7747028","level":"debug","message":"the quota has not changed","quota":200000,"serviceContext":{"service":"ws-daemon","version":"commit-b7b4c81e4c98efecca85740b0e07a2ebf79261f8"},"severity":"DEBUG","time":"2022-01-25T00:21:39Z","userId":"","workspaceId":"gitpodio-templatetypescr-loh4kwdzby1"}

```

## Release Notes
```release-note
[ws-daemon] Refactor enforcing of CPU limits
```

## Documentation
